docs/source/en/tasks/knowledge_distillation_for_image_classification.md

```diff
@@ -0,0 +1,183 @@
+
+Let's install the libraries needed for distillation and evaluating the process. 
+
+```bash
+pip install transformers datasets accelerate tensorboard evaluate --upgrade
+```
+
+In this example, we are using `microsoft/resnet-50`, trained with ImageNet-1k dataset on resolution of 224x224. You can see that feature extractor returns the same output below.
+
+```python
+from transformers import AutoFeatureExtractor
+from PIL import Image
+import requests
+import numpy as np
+
+teacher_extractor = AutoFeatureExtractor.from_pretrained("microsoft/resnet-50")
+url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+sample = Image.open(requests.get(url, stream=True).raw)
+```
+
+We will now load and pre-process the dataset. 
+
+```python
+from datasets import load_dataset
+
+dataset = load_dataset("beans")
+```
+
+We can use either of the processors given they return the same output. We will use `map()` method of `dataset` to apply the preprocessing to every split of the dataset.
+
+```python
+def process(examples):
+    processed_inputs = teacher_extractor(examples["image"])
+    return processed_inputs
+
+processed_datasets = dataset.map(process, batched=True)
+```
+
+Essentially, we want the student to mimick the teacher so we will get the output from the teacher and the student, multiply it with a parameter called `temperature`, then calculate KL-divergence between them. We will then calculate loss based on KL-divergence and student output, and to provide a trade-off between them, we will use a parameter called `lambda`. In this tutorial, we are going to use `lambda` as 0.5 and temperature as 5.
+
+
+```python
+from transformers import TrainingArguments, Trainer
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class ImageDistilTrainer(Trainer):
+    def __init__(self, *args, teacher_model=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.teacher = teacher_model
+        self.student = student_model
+        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+        self.teacher.to(device)
+        self.teacher.eval()
+
+    def compute_loss(self, student, inputs, return_outputs=False):
+        lambda_param = 0.5
+        temperature = 5
+        student_output = self.student(**inputs)
+
+        with torch.no_grad():
+          teacher_output = self.teacher(**inputs)
+
+        loss_function = nn.KLDivLoss(reduction="batchmean")
+        loss_logits = (loss_function(
+            F.log_softmax(student_output.logits / temperature, dim=-1),
+            F.softmax(teacher_output.logits / temperature, dim=-1)) * (temperature ** 2))
+
+        loss = lambda_param * student_output.loss + (1. - lambda_param) * loss_logits
+        return (loss, student_output) if return_outputs else loss
```

Comment:
Adding comments, renaming some variables and separating the soft targets:

```python
    def compute_loss(self, _, inputs, return_outputs=False):
        student_output = self.student(**inputs)

        with torch.no_grad():
            teacher_output = self.teacher(**inputs)

        # Compute each soft target
        soft_teacher = F.softmax(teacher_output.logits / self.temperature, dim=-1)
        soft_student = F.log_softmax(student_output.logits / self.temperature, dim=-1)
        # Compute the distillation loss
        distillation_loss = self.loss_function(soft_student, soft_teacher) * (self.temperature ** 2)
        # Compute the true label loss (students loss)
        student_target_loss = student_output.loss
        # Weight the two losses together
        loss = (1. - self.lambda_param) * student_target_loss + self.lambda_param * distillation_loss
        return (loss, student_output) if return_outputs else loss
```

Reply:
That's nice, I wanted to keep it minimal but this would be better I guess.